File: 2022-2023/sourcecode/lab/lab11/OkHttpLoggingInterceptorBasicsJava/app/src/main/java/pl/udu/uwr/pum/okhttplogginginterceptorbasicsjava/RetrofitFactory.java
package pl.udu.uwr.pum.okhttplogginginterceptorbasicsjava;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public final class RetrofitFactory {
    private RetrofitFactory(){}

    private static volatile PlaceholderService INSTANCE;
    private final String url = "https://jsonplaceholder.typicode.com/";

    public static PlaceholderService makeService() {
        if (INSTANCE == null) {
            synchronized (RetrofitFactory.class) {
                if (INSTANCE == null) {
                    HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
                    interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
                    OkHttpClient client = new OkHttpClient.Builder()
                            .addInterceptor(interceptor)
                            .build();

                    INSTANCE = new Retrofit.Builder()
                            .baseUrl("https://jsonplaceholder.typicode.com/")
                            .addConverterFactory(GsonConverterFactory.create())
                            .client(client)
                            .build().create(PlaceholderService.class);
                }
            }
        }
        return INSTANCE;
    }
}

File: 2022-2023/sourcecode/lab/lab11/OkHttpLoggingInterceptorBasicsKotlin/app/src/main/java/pl/udu/uwr/pum/okhttplogginginterceptorbasicskotlin/RetrofitFactory.kt
package pl.udu.uwr.pum.okhttplogginginterceptorbasicskotlin

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitFactory {

    private const val url = "https://jsonplaceholder.typicode.com/"

    val service: PlaceholderService by lazy {
        val interceptor = HttpLoggingInterceptor()
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val client = OkHttpClient.Builder()
            .addInterceptor(interceptor)
            .build()

        Retrofit.Builder()
            .baseUrl(url)
            .addConverterFactory(GsonConverterFactory.create())
            .client(client)
            .build().create(PlaceholderService::class.java)
    }
}